Build NuGet package (#281)
signalfx/splunk-otel-dotnet: src/Splunk.OpenTelemetry.AutoInstrumentation/nuget-readme.md

# Using the Splunk.OpenTelemetry.AutoInstrumentation NuGet package

> ## When to use the NuGet package

Author: It is new for us, but, we will actively support it. However, at least for the time being, the "deployment" mode should be the default, and the NuGet should be used/suggested in very specific cases. We did have some cases in which users saw the deployment instructions and asked: "don't you have a NuGet package instead?"

Use the NuGet package in the following scenarios:

> 1. Simplify deployment. For example, a container running a single application.

Reviewer: In general, I like the idea to have 1. for all points, but this document can be displayed also on nuget.org, Visual Studio, possibly other nuget managers/viewers, and also in readme after upacking nuget. Please make it readable without special parsing. In this case use real order.

Author: For reference https://learn.microsoft.com/en-us/nuget/nuget-org/package-readme-on-nuget-org#supported-markdown-features

Author: Just to be clear: it will be good if the tool has some markdown lint targeted to the features supported on nuget.org - the repo already has markdown lint.

> 1. Simplify deployment. For example, a container running a single application.

Reviewer 2: "Simplify deployment" is a bit subjective. Why wouldn't a developer want to use this instead of the installer MSI or script? Consider removing or rewriting this unless NuGet packages are meant to be the primary method going forward.

Author: Agreed, this is subjective. The objective case here is when the developer doesn't have control over the computer image where the application is running, example: developer ships the app but other team is in charge of the container. For such cases the NuGet allow the developer to ship the instrumentation.

Some also consider it more "convenient" to do it via the NuGet package, however, that is subjective too.

I will try to re-word it to avoid the subjective aspect.

Author: 04122caa2835825cf789537c070d852f73cb0e22

1. Support instrumentation of [`self-contained`](https://learn.microsoft.com/en-us/dotnet/core/deploying/#publish-self-contained)
  applications.
1. Facilitate developer experimentation with automatic instrumentation through
  NuGet packages.
1. Solve version conflicts between the dependencies used by the application and the
  automatic instrumentation.

## Limitations

While NuGet packages are a convenient way to deploy automatic
instrumentation, they can't be used in all cases. The most common
reasons for not using NuGet packages include the following:

1. You can't add the package to the application project. For example,
the application is from a third party that can't add the package.
1. Reduce disk usage, or the size of a virtual machine, when multiple applications
to be instrumented are installed in a single machine. In this case you can use
a single deployment for all .NET applications running on the machine.

> 1. A legacy application that can't be migrated to the [SDK-style project](https://learn.microsoft.com/en-us/nuget/resources/check-project-format#check-the-project-format).

Reviewer 2: How common are these scenarios among our users? If not that much, it promotes the idea of turning this method into the main one... Otherwise, we'll need stronger reasons for not using NuGets in all cases.

Author: The most common case here are ASP.NET apps. So they are all on Windows and typically with IIS, for which our deployment via PowerShell/MSI is preferrable since typically there are multiple IIS apps in the same host.

## Using the NuGet packages

To automatically instrument your application with the Splunk Distribution
of OpenTelemetry .NET add
the `Splunk.OpenTelemetry.AutoInstrumentation` package to your project:

```terminal
dotnet add [<PROJECT>] package Splunk.OpenTelemetry.AutoInstrumentation --prerelease
```

> If the application references packages that can be instrumented, but, require

Reviewer 2: ```suggestion
If the application references packages that can be instrumented, but require
```

> other packages for the instrumentation to work the build will fail and prompt

Reviewer 2: ```suggestion
other packages for the instrumentation to work, the build will fail and prompt
```

you to either add the missing instrumentation package or to skip the
instrumentation of the corresponding package:

```terminal
~packages/opentelemetry.autoinstrumentation.buildtasks/1.0.0-rc.1/build/OpenTelemetry.AutoInstrumentation.BuildTasks.targets(29,5): error : OpenTelemetry.AutoInstrumentation: add a reference to the instrumentation package 'MongoDB.Driver.Core.Extensions.DiagnosticSources' version 1.3.0 or add 'MongoDB.Driver.Core' to the property 'SkippedInstrumentations' to suppress this error.
```

To resolve the error either add the recommended instrumentation package or skip
the instrumentation of the listed package by adding it to the `SkippedInstrumentation`

> property, example:

Reviewer 2: ```suggestion
property. For example:
```

Author: c5448afec7b0d390fd129e4de772433d59d78c46

```csproj
  <PropertyGroup>
    <SkippedInstrumentations>MongoDB.Driver.Core;StackExchange.Redis</SkippedInstrumentations>
  </PropertyGroup>
```

> The same property can be also specified directly via the CLI, notice that the

Reviewer 2: ```suggestion
The same property can be also specified directly using the terminal. Notice that the
```

> separator, `;`, needs to be properly escaped as '%3B':

Reviewer 2: ```suggestion
`;` separator needs to be properly escaped as '%3B':
```
